ComponentKit/Core/CKComponentViewAttribute.h

```diff
@@ -74,52 +113,216 @@ struct CKComponentViewAttribute {
-  CKBoxedValue(float v) noexcept : __actual(@(v)) {};
-  CKBoxedValue(double v) noexcept : __actual(@(v)) {};
-  CKBoxedValue(SEL v) noexcept : __actual([NSValue valueWithPointer:v]) {};
-  CKBoxedValue(std::nullptr_t v) noexcept : __actual(nil) {};
-
-  // Any objects go here
-  CKBoxedValue(id obj) noexcept : __actual(obj) {};
-
-  // Define conversions for common Apple types
-  CKBoxedValue(CGRect v) noexcept : __actual([NSValue valueWithCGRect:v]) {};
-  CKBoxedValue(CGPoint v) noexcept : __actual([NSValue valueWithCGPoint:v]) {};
-  CKBoxedValue(CGSize v) noexcept : __actual([NSValue valueWithCGSize:v]) {};
-  CKBoxedValue(UIEdgeInsets v) noexcept : __actual([NSValue valueWithUIEdgeInsets:v]) {};
-
-  operator id () const {
-    return __actual;
-  };
+namespace CK {
+  namespace ViewAttribute {
 
-private:
-  id __actual;
+    struct CachedSetter {
+      NSInvocation *invocation;
+      NSUInteger argumentSize;
+      const char *argumentType;
 
-};
+      CachedSetter(NSInvocation *inv, NSUInteger argSize, const char *argType) :
+      invocation(inv),
+      argumentSize(argSize),
+      argumentType(argType) {}
+    };
+
+    extern const CachedSetter &CachedSetterInvocation(id object, SEL setter);
+
+    // Singleton object to signify non-ID-type object. For internal use only.
+    extern id nonIDObject;
+
+    /**
+     An abstract base type for the value type. This is a hack to store multiple templated objects in a non
+     templated collection.
+     */
+    struct ValueBase
+    {
+      ValueBase(std::string typeName) : _typeName(typeName) {}
+      ValueBase(const ValueBase &) = delete;
+      ~ValueBase() {}
+
+      virtual operator id() const = 0;
+      virtual size_t hash() const noexcept = 0;
+      virtual BOOL isEqualTo(const ValueBase &) const = 0;
+      virtual void performSetter(id object, SEL setter) const = 0;
+
+      const std::string typeName() const
+      {
+        return _typeName;
+      }
+
+    protected:
+      const std::string _typeName;
+    };
+
+    /**
+     Templated type used to store an underlying value.
+     */
+    template<typename T> struct Value : ValueBase
+    {
+      Value(const T &value) : ValueBase(typeid(Value<T>).name()), _value(value) {}
+      Value(const Value<T> &) = delete;
+      ~Value() {}
+
+      operator T() const
+      {
+        return _value;
+      }
+
+      operator id() const
+      {
+        return nonIDObject;
+      }
+
+      size_t hash() const noexcept
+      {
+        return CK::hash<T>()(_value);
+      }
+
+      BOOL isEqualTo(const ValueBase &other) const
+      {
+        return this->typeName() == other.typeName() && CK::is_equal<T>()(_value, static_cast<const Value<T> &>(other)._value);
+      }
+
+      void performSetter(id object, SEL setter) const
+      {
+        const CachedSetter &set = CachedSetterInvocation(object, setter);
+        T value = _value;
+        [set.invocation setArgument:&value atIndex:2];
```

Comment:
If you're using an invocation then the type should be trivially-constructible, right? What happens if I pass a std::vector<CGRect> here.